### PR TITLE
Use configurationRedirection for IIS Deployer

### PR DESCRIPTION
**Pull request by pakrym**
https://github.com/aspnet/IISIntegration/issues/1169

### CODE REVIEW

File: src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISDeployer.cs [hunk 1]
```diff
@@ -100,28 +110,247 @@ public override async Task<DeploymentResult> DeployAsync()
...
+                        if (site.State == ObjectState.Stopped)
+                        {
+                            if (HostProcess.HasExited)
+                            {
+                                Logger.LogInformation($"Site has stopped successfully.");
+                                return;
+                            }
+                        }
+                    }
+                    catch (COMException)
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISDeployer.cs]: This should probably catch the same exceptions as `WaitUntilSiteStarted`.

File: src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISDeployer.cs [hunk 2]
```diff
@@ -100,28 +110,247 @@ public override async Task<DeploymentResult> DeployAsync()
...
+
+                serverManager.CommitChanges();
+
+                Directory.Delete(_configPath, true);
+            }
+        }
+
+        private void ConfigureAppHostConfig(XElement config, string contentRoot, int port)
+        {
+            var site = config
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISDeployer.cs]: Nit: formatting in this method is inconsistent.

File: src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISDeployerBase.cs [hunk 3]
```diff
@@ -26,49 +28,62 @@ public IISDeployerBase(IISDeploymentParameters deploymentParameters, ILoggerFact
...
             {
-                action.Invoke(xElement, contentRoot);
+                action.Invoke(webconfig.Root, contentRoot);
             }
 
-            foreach (var action in IISDeploymentParameters.WebConfigActionList)
+            if (IISDeploymentParameters != null)
             {
-                action.Invoke(xElement, contentRoot);
+                foreach (var action in IISDeploymentParameters.WebConfigActionList)
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISDeployerBase.cs]: Nit: `IISDeploymentParameters?.WebConfigActionList`

File: src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/XElementExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+            var existing = element.Element(name);
+            if (existing == null)
+            {
+                throw new InvalidOperationException($"Element with name {name} not found in {element}");
+            }
+
+            return existing;
+        }
+
+        public static XElement AddOrUpdate(this XElement element, string name)
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/XElementExtensions.cs]: Nit: this doesn't update the existing element.

Comment (pakrym) [reply src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/XElementExtensions.cs]: GetOrAdd?

File: src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISDeployerBase.cs (hunk 3 above)
Comment (pakrym) [reply src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISDeployerBase.cs]: You can't `foreach` null

File: src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISDeployer.cs [hunk 5]
```diff
@@ -100,28 +110,247 @@ public override async Task<DeploymentResult> DeployAsync()
...
+                    await Task.Delay(_retryDelay);
+                }
+
+                throw new TimeoutException($"IIS failed to stop site {site}.");
+            }
+        }
+
+        private void AddTemporaryAppHostConfig(string contentRoot, int port)
+        {
+            _configPath = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString("N"));
```
[the remark this answers is not in the document]
Comment (pakrym) [reply src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISDeployer.cs]: No, because some tests delete publish output to check that it's not locked

**Comment by pakrym** [comment]
:up::date:

File: src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISDeployer.cs [hunk 6]
```diff
@@ -244,7 +241,7 @@ private async Task WaitUntilSiteStopped()
                             }
                         }
                     }
-                    catch (COMException)
+                    catch (COMException ex) when (IsExpectedException(ex))
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISDeployer.cs]: catch `Exception`

**Pull request merged**